src/esrs/store.rs

```diff
@@ -1,55 +1,41 @@
-use serde::de::DeserializeOwned;
-use serde::Serialize;
 use uuid::Uuid;
 
-use crate::esrs::SequenceNumber;
+use crate::types::SequenceNumber;
+use crate::{Aggregate, AggregateManager};
 
 /// An EventStore is responsible for persisting events that an aggregate emits into a database, and loading the events
 /// that represent an aggregate's history from the database.
 #[async_trait]
-pub trait EventStore<Event: Serialize + DeserializeOwned + Send + Sync, Error> {
+pub trait EventStore {
+    type Manager: AggregateManager;
+
     /// Loads the events that an aggregate instance has emitted in the past.
-    async fn by_aggregate_id(&self, id: Uuid) -> Result<Vec<StoreEvent<Event>>, Error>;
+    async fn by_aggregate_id(
+        &self,
+        aggregate_id: Uuid,
+    ) -> Result<Vec<StoreEvent<<Self::Manager as Aggregate>::Event>>, <Self::Manager as Aggregate>::Error>;
 
-    /// Persists multiple events into the database.  This should be done in a single transaction - either
+    /// Persists multiple events into the database. This should be done in a single transaction - either
     /// all the events are persisted correctly, or none are.
     ///
     /// Persisting events may additionally trigger configured Projectors.
     async fn persist(
         &self,
         aggregate_id: Uuid,
-        events: Vec<Event>,
+        events: Vec<<Self::Manager as Aggregate>::Event>,
         starting_sequence_number: SequenceNumber,
-    ) -> Result<Vec<StoreEvent<Event>>, Error>;
-
-    /// Run any policies attached to this store against a set of events.
-    /// This should be called only after the events have successfully been persisted in the store.
-    async fn run_policies(&self, events: &[StoreEvent<Event>]) -> Result<(), Error>;
-
-    async fn close(&self);
+    ) -> Result<Vec<StoreEvent<<Self::Manager as Aggregate>::Event>>, <Self::Manager as Aggregate>::Error>;
 
-    fn get_all(&self) -> BoxStream<Result<StoreEvent<Event>, Error>>;
-}
-
-/// A ProjectorStore is responsible for projecting an event (that has been persisted to the database) into a
-/// form that is better suited to being read by other parts of the application.
-#[async_trait]
-pub trait ProjectorStore<Event: Serialize + DeserializeOwned + Send + Sync, Executor, Error> {
-    async fn project_event(&self, store_event: &StoreEvent<Event>, executor: &mut Executor) -> Result<(), Error>;
-}
-
-/// An EraserStore is responsible for wiping an aggregate instance from history: it should delete the
-/// aggregate instance, along with all of its events, or fail.
-#[async_trait]
-pub trait EraserStore<Event: Serialize + DeserializeOwned + Send + Sync, Error> {
-    async fn delete(&self, aggregate_id: Uuid) -> Result<(), Error>;
+    /// Delete all events from events store related to given `aggregate_id`.
+    ///
+    /// Moreover it should delete all the projections.
+    async fn delete_by_aggregate_id(&self, aggregate_id: Uuid) -> Result<(), <Self::Manager as Aggregate>::Error>;
```

Comment:
Thoughts about calling the methods of this trait `load` , `persist` and `delete`? It's kind of clear that we load and delete by id

Reply:
Yeah, ok, i'll change it to delete

Reply:
About the load i don't agree because it actually do not load anything but just returns a list of events..